src/main/java/gov/nasa/pds/api/engineering/controllers/MyBundlesApiController.java

```diff
@@ -196,8 +190,122 @@ private Products getCollectionChildren(String lidvid, int start, int limit, List
+		 String accept = this.request.getHeader("Accept");
+		 MyBundlesApiController.log.info("accept value is " + accept);
+		 if ((accept != null 
+		 		&& (accept.contains("application/json") 
+		 				|| accept.contains("text/html")
+		 				|| accept.contains("application/xml")
+		 				|| accept.contains("application/pds4+xml")
+		 				|| accept.contains("*/*")))
+		 	|| (accept == null)) {
+		 	
+		 	try {
+		    	
+		 	
+		 		Products products = this.getProductChildren(lidvid, start, limit, fields, sort, onlySummary);
+		    	
+		    	return new ResponseEntity<Products>(products, HttpStatus.OK);
+		    	
+		  } catch (IOException e) {
+		       log.error("Couldn't serialize response for content type " + accept, e);
+		       return new ResponseEntity<Products>(HttpStatus.INTERNAL_SERVER_ERROR);
+		  }
+		     
+		 }
+		 else return new ResponseEntity<Products>(HttpStatus.NOT_IMPLEMENTED);
+	}
+
+    @SuppressWarnings("unchecked")
+	private Products getProductChildren(String lidvid, int start, int limit, List<String> fields, List<String> sort, boolean onlySummary) throws IOException
+    {
+		if (!lidvid.contains("::") && !lidvid.endsWith(":")) lidvid = this.getLatestLidVidFromLid(lidvid);
+    	MyBundlesApiController.log.info("request bundle lidvid, children of products: " + lidvid);
+       	
+    	GetRequest getBundleRequest = new GetRequest(this.esRegistryConnection.getRegistryIndex(), lidvid);
+    	HashSet<String> uniqueProperties = new HashSet<String>();
+    	List<String> productLidvids = new ArrayList<String>();
+    	Products products = new Products();
+        RestHighLevelClient restHighLevelClient = this.esRegistryConnection.getRestHighLevelClient();
+      	Summary summary = new Summary();
+
+    	if (sort == null) { sort = Arrays.asList(); }
+
+    	summary.setStart(start);
+    	summary.setLimit(limit);
+    	summary.setSort(sort);
+    	products.setSummary(summary);
+
+    	try
+    	{
+    		GetResponse getBundleResponse = restHighLevelClient.get(getBundleRequest, RequestOptions.DEFAULT);
+
+	    	if (getBundleResponse.isExists())
+	    	{
+	    		MyBundlesApiController.log.info("get response " + getBundleResponse.toString());
+				List<String> collections = (List<String>) getBundleResponse.getSourceAsMap().get("ref_lid_collection");
+        		String collectionLidVid;
+
+        		for (String collectionLid : collections )
+        		{
+        			collectionLidVid = this.getLatestLidVidFromLid(collectionLid) + "::P1";
+        			MyBundlesApiController.log.info("get collection with lidvid " + collectionLidVid);
+        			GetRequest getCollectionRequest = new GetRequest(this.esRegistryConnection.getRegistryRefIndex(), collectionLidVid);
+        			GetResponse getCollectionResponse = restHighLevelClient.get(getCollectionRequest, RequestOptions.DEFAULT);
+
+        			if (getCollectionResponse.isExists())
+        			{
+        				MyBundlesApiController.log.info("get ref response " + getCollectionResponse.toString());
```

Comment:
The iteratable class available on https://github.com/NASA-PDS/registry-api-service/tree/master/src/main/java/gov/nasa/pds/api/engineering/elasticsearch/business can be used to find the list of products of a collection between start and start+limit without having to load all products in memory.

This method could also be implemented as an iterator.

That will be convenient to re-use the code since we want also to optimize the code which access the products of the collection by doing a single request for all the products instead of one per product as you do in line 277.

Since that works anyway here, we can work on the code re-useability in the optimization ticket https://github.com/nasa-pds/registry-api-service/issues/13

Reply:
It cannot. In order to complete the search, first have to get the bundle then all collections then all products. The limits and pagination are over all products not bundles and collections, therefore bundle and collections have to be read to build the products list before pagination can take place. Ugly, but that is part of one of the various emails with @jordanpadams were the punch line was pagination is products not collections in this case.

As I stated in that email also, the iterator also reads all the data. It is not possible to do fractional searches unless state is saved and all these interfaces are REST.